arrow/obsidian.c

```diff
@@ -0,0 +1,279 @@
+//   * steps: How many steps were performed
+//   * time: An array of length `steps` containing the value at each time point
+//   * events: An array of length `steps` signifying which reaction took place at
+//       each time point
+//   * outcome: The final state after all of the reactions have been performed.
+evolve_result
+evolve(int reactions_count,
+       int substrates_count,
+       long * stoichiometry,
+       double * rates,
+
+       long * reactants_lengths,
+       long * reactants_indexes,
+       long * reactants,
+       long * reactions,
+       
+       long * dependencies_lengths,
+       long * dependencies_indexes,
+       long * dependencies,
+
+       long * substrates_lengths,
+       long * substrates_indexes,
+       long * substrates,
+
+       double duration,
+       long * state) {
+
+  // The `event_bounds` will be used to determine how much space to allocate for
+  // tracking the evolution of the system's state. If a step is reached that exceeds
+  // `event_bounds` then these arrays will be reallocated after doubling `event_bounds`.
+  long event_bounds = INITIAL_LENGTH;
+
+  // Allocate the dynamic arrays that will be used to track the progress of the system
+  double * time = malloc((sizeof (double *)) * event_bounds);
+  long * events = malloc((sizeof (long *)) * event_bounds);
+  long * outcome = malloc((sizeof (long *)) * substrates_count);
+
+  // Allocate space for the temporary values that will be used entirely within this
+  // function
+  double * propensities = malloc((sizeof (double *)) * reactions_count);
+  long * update = malloc((sizeof (long *)) * reactions_count);
+  long update_length = reactions_count;
+
+  // Declare the working variables we will use throughout this function
+  long substrates_length;
+  long reaction, reactant, species, index, involve, count, adjustment;
+  double total, interval, sample, progress, point;
+  int choice, step = 0, up = 0;
+  double now = 0.0;
+
+  // Copy the initial state that was supplied from outside to the working `outcome`
+  // array we will use to actually apply the reactions and determine the next step's
+  // state.
+  for (species = 0; species < substrates_count; species++) {
+    outcome[species] = state[species];
+  }
+
+  // The `update` array will hold for each step which propensities need to be updated
+  // for the next time step, based on the dependencies between reactions (sharing
+  // substrates).
+  for (reaction = 0; reaction < reactions_count; reaction++) {
+    update[reaction] = reaction;
+  }
+
+  // Calculate steps until we reach the provided duration
+  while(now < duration) {
+
+    // First update all propensities that were affected by the previous event
+    for (up = 0; up < update_length; up++) {
+
+      // Find which reaction to update and initialize the propensity for that reaction
+      // with the rate for that reaction
+      reaction = update[up];
+      propensities[reaction] = rates[reaction];
```

Comment:
Also, in all the cases `reaction` is used it refers to the *same index*, so seems apropos to me.

Reply:
Well, I totally missed that assignment to `reaction` on the previous line!

BTW I'm not so picky about `while (...)` and `sizeof(...)` as with the `*`s. As you noted, pointers are very significant and easy to goof. (At least we don't have the nasty case where it's not obvious what the code is incrementing when it has a pointer to an array of image bitplanes, which are arrays of scanlines, which are arrays of pixels.)

FYI, it's common to write a loop increment like `p++` but I don't know why. The postfix increment operator is a strange beast "return the variable's contents and by the way side-effect increment the variable." Personally I think it's better practice to use it only when needed, like `*p++` (which is a PDP-11 addressing mode), otherwise use `++p` which straightforwardly says "increment this variable". But it's OK either way.